src/nnet3/nnet-discriminative-example.cc

```diff
@@ -0,0 +1,419 @@
+        deriv_weights(t * num_sequences + s) = 0.0;
+  }
+}
+
+void GetDiscriminativeComputationRequest(const Nnet &nnet,
+                                         const NnetDiscriminativeExample &eg,
+                                         bool need_model_derivative,
+                                         bool store_component_stats,
+                                         bool use_xent_regularization,
+                                         bool use_xent_derivative,
+                                         ComputationRequest *request) {
+  request->inputs.clear();
+  request->inputs.reserve(eg.inputs.size());
+  request->outputs.clear();
+  request->outputs.reserve(eg.outputs.size());
+  request->need_model_derivative = need_model_derivative;
+  request->store_component_stats = store_component_stats;
+  for (size_t i = 0; i < eg.inputs.size(); i++) {
+    const NnetIo &io = eg.inputs[i];
+    const std::string &name = io.name;
+    int32 node_index = nnet.GetNodeIndex(name);
+    if (node_index == -1 &&
+        !nnet.IsInputNode(node_index))
+      KALDI_ERR << "Nnet example has input named '" << name
+                << "', but no such input node is in the network.";
+
+    request->inputs.resize(request->inputs.size() + 1);
+    IoSpecification &io_spec = request->inputs.back();
+    io_spec.name = name;
+    io_spec.indexes = io.indexes;
+    io_spec.has_deriv = false;
+  }
+  for (size_t i = 0; i < eg.outputs.size(); i++) {
+    // there will normally be exactly one output , named "output"
+    const NnetDiscriminativeSupervision &sup = eg.outputs[i];
+    const std::string &name = sup.name;
+    int32 node_index = nnet.GetNodeIndex(name);
+    if (node_index == -1 &&
+        !nnet.IsOutputNode(node_index))
+      KALDI_ERR << "Nnet example has output named '" << name
+                << "', but no such output node is in the network.";
+    request->outputs.resize(request->outputs.size() + 1);
+    IoSpecification &io_spec = request->outputs.back();
+    io_spec.name = name;
+    io_spec.indexes = sup.indexes;
+    io_spec.has_deriv = need_model_derivative;
+    
+    if (use_xent_regularization) {
+      size_t cur_size = request->outputs.size();
+      request->outputs.resize(cur_size + 1);
+      IoSpecification &io_spec = request->outputs[cur_size - 1],
+          &io_spec_xent = request->outputs[cur_size];
+      // the IoSpecification for the -xent output is the same
+      // as for the regular output, except for its name which has
+      // the -xent suffix (and the has_deriv member may differ).
+      io_spec_xent = io_spec;
+      io_spec_xent.name = name + "-xent";
+      io_spec_xent.has_deriv = use_xent_derivative;
+    }
+  }
+  // check to see if something went wrong.
+  if (request->inputs.empty())
+    KALDI_ERR << "No inputs in computation request.";
+  if (request->outputs.empty())
+    KALDI_ERR << "No outputs in computation request.";
+}
+
+void ShiftDiscriminativeExampleTimes(int32 frame_shift,
```

Comment:
(a) is what I have results for. I can make the scripts to support (b) to
try later.

On Tue, Mar 8, 2016, 20:43 Daniel Povey <notifications@github.com> wrote:

> In src/nnet3/nnet-discriminative-example.cc
> <https://github.com/kaldi-asr/kaldi/pull/563#discussion_r55460891>:
>
> > +      // the IoSpecification for the -xent output is the same
> > +      // as for the regular output, except for its name which has
> > +      // the -xent suffix (and the has_deriv member may differ).
> > +      io_spec_xent = io_spec;
> > +      io_spec_xent.name = name + "-xent";
> > +      io_spec_xent.has_deriv = use_xent_derivative;
> > +    }
> > +  }
> > +  // check to see if something went wrong.
> > +  if (request->inputs.empty())
> > +    KALDI_ERR << "No inputs in computation request.";
> > +  if (request->outputs.empty())
> > +    KALDI_ERR << "No outputs in computation request.";
> > +}
> > +
> > +void ShiftDiscriminativeExampleTimes(int32 frame_shift,
>
> Actually, I think it would make sense to compare 2 ways of doing the
> shifts: - (a), dumping the lattices with no shift and doing the shift at
> the egs level the same way we do for regular chain training, with the
> 'ShiftDiscriminativeExample()' function or whatever it was called, - (b), a
> more exact method where we dump the lattices at all shifts. This is just
> for doing due diligence on the chain+discriminative-training setup. I don't
> really think it's going to give us much improvement, and it's not that
> urgent. Dan
> … <#msg-f:1528287055664639253_msg-f:1528286726484747763_>
> On Tue, Mar 8, 2016 at 8:32 PM, Daniel Povey <dpovey@gmail.com> wrote: >
> No, I am saying you should not shift without regenerating the lattices >
> with the data at the correct shift. It would be easiest for now to do it >
> using shift-feats at the command line level, and just dump separate egs. >
> Dan > > > On Tue, Mar 8, 2016 at 7:53 PM, Vimal Manohar <
> notifications@github.com> > wrote: > In
> src/nnet3/nnet-discriminative-example.cc > <#563 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/563#discussion_r55456425>>: > >
> > + // the IoSpecification for the -xent output is the same > > + // as for
> the regular output, except for its name which has > > + // the -xent suffix
> (and the has_deriv member may differ). > > + io_spec_xent = io_spec; > > +
> io_spec_xent.name = name + "-xent"; > > + io_spec_xent.has_deriv =
> use_xent_derivative; > > + } > > + } > > + // check to see if something
> went wrong. > > + if (request->inputs.empty()) > > + KALDI_ERR << "No
> inputs in computation request."; > > + if (request->outputs.empty()) > > +
> KALDI_ERR << "No outputs in computation request."; > > +} > > + > > +void
> ShiftDiscriminativeExampleTimes(int32 frame_shift, > > For chain models, I
> have sub-sampled alignments and lattices. So I need > to shift input by 1
> and 2 frames in the decode and align scripts to get the > different shifts?
> > Also would this affect the chain + sMBR results? > > — > Reply to this
> email directly or view it on GitHub > <
> https://github.com/kaldi-asr/kaldi/pull/563/files#r55456425>. >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/563/files#r55460891>.
>
-- 
Vimal Manohar
PhD Student
Electrical & Computer Engineering
Johns Hopkins University